### PR TITLE
[Paddle-TRT] Implement MHA fp16 order same as training

### PR DESCRIPTION
**Pull request by zlsh80826**
### PR types

Others

### PR changes

OPs

### Describe

[DEMO ONLY]
TRT MHA plugin change to apply scale before QK multiplication

### CODE REVIEW

File: paddle/fluid/inference/tensorrt/plugin/qkv_to_context_plugin.cu [hunk 1]
```diff
@@ -225,6 +225,12 @@ nvinfer1::DataType QkvToContextPluginDynamic::getOutputDataType(
   return input_types[0];
 }
 
+template <typename T>
```
Comment (shangzhizhou) [inline paddle/fluid/inference/tensorrt/plugin/qkv_to_context_plugin.cu:228]: 该kernel需要判断__cuda_arch__，否则在不支持fp16的cuda arch上会编译失败。

```
template <>
__global__ void apply_scale(half* data, half scale, int n) {
#if CUDA_ARCH_FP16_SUPPORTED(__CUDA_ARCH__)
  int tid = blockIdx.x * blockDim.x + threadIdx.x;
  data[tid] = data[tid] * scale;
#endif
}
```

**Review by shangzhizhou — DISMISSED** [review]
LGTM

File: paddle/fluid/inference/tensorrt/plugin/qkv_to_context_plugin.cu [hunk 2]
```diff
@@ -225,6 +225,14 @@ nvinfer1::DataType QkvToContextPluginDynamic::getOutputDataType(
...
+__global__ void apply_scale(T* data, T scale, int n) {
```
Comment (shangzhizhou) [inline paddle/fluid/inference/tensorrt/plugin/qkv_to_context_plugin.cu]: format error，是不是pre-commit没有配置好？
```
-__global__ void apply_scale(T* data, T scale, int n) {
+__global__ void apply_scale(T *data, T scale, int n) {
```

File: paddle/fluid/inference/tensorrt/plugin/qkv_to_context_plugin.cu [hunk 3]
```diff
@@ -291,10 +299,16 @@ int QkvToContextPluginDynamic::enqueue(
         platform::DeviceContextPool::Instance().Get(
             platform::CUDAPlace(device_id)));
 
+    int n_q = seq_len * head_number_ * head_size_;
+    constexpr int threads = 128;
+    int blocks = (n_q + threads - 1) / threads;
+
+    apply_scale<<<blocks, threads, 0, stream>>>(tptr, static_cast<half>(scale_), n_q);
```
Comment (shangzhizhou) [inline paddle/fluid/inference/tensorrt/plugin/qkv_to_context_plugin.cu]: format error.
```
-    apply_scale<<<blocks, threads, 0, stream>>>(tptr, static_cast<half>(scale_), n_q);
+    apply_scale<<<blocks, threads, 0, stream>>>(tptr, static_cast<half>(scale_),
+                                                n_q); 
```

File: paddle/fluid/inference/tensorrt/plugin/qkv_to_context_plugin.cu (hunk 2 above)
Comment (zlsh80826) [reply paddle/fluid/inference/tensorrt/plugin/qkv_to_context_plugin.cu]: DONE

**Review by shangzhizhou — APPROVED** [review]
LGTM

**Pull request merged**